pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,246 @@
+	"strings"
+	"sync"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	"golang.org/x/sys/unix"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdrop/tracer/tcpbits"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdrop/types"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -cc clang -no-global-types -type event tcpdrop ./bpf/tcpdrop.bpf.c -- -I./bpf/ -I../../../../${TARGET}
+
+type Config struct{}
+
+type Tracer struct {
+	config        *Config
+	enricher      gadgets.DataEnricherByMntNs
+	eventCallback func(*types.Event)
+
+	objs tcpdropObjects
+
+	kfreeSkbLink link.Link
+
+	reader *perf.Reader
+}
+
+func NewTracer(config *Config, enricher gadgets.DataEnricherByMntNs,
+	eventCallback func(*types.Event),
+) (*Tracer, error) {
+	t := &Tracer{
+		config:        config,
+		enricher:      enricher,
+		eventCallback: eventCallback,
+	}
+
+	if err := t.install(); err != nil {
+		t.close()
+		return nil, err
+	}
+
+	go t.run()
+
+	return t, nil
+}
+
+func (t *Tracer) close() {
+	t.kfreeSkbLink = gadgets.CloseLink(t.kfreeSkbLink)
+
+	if t.reader != nil {
+		t.reader.Close()
+	}
+
+	t.objs.Close()
+}
+
+func (t *Tracer) install() error {
+	spec, err := loadTcpdrop()
+	if err != nil {
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		var errVerifier *ebpf.VerifierError
+		if errors.As(err, &errVerifier) {
+			fmt.Printf("Verifier error: %+v\n",
+				errVerifier)
+		}
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	t.kfreeSkbLink, err = link.Tracepoint("skb", "kfree_skb", t.objs.IgTcpdrop, nil)
+	if err != nil {
+		return fmt.Errorf("error opening tracepoint kfree_skb: %w", err)
+	}
+
+	reader, err := perf.NewReader(t.objs.tcpdropMaps.Events, gadgets.PerfBufferPages*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("error creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if record.LostSamples > 0 {
+			msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+			t.eventCallback(types.Base(eventtypes.Warn(msg)))
+			continue
+		}
+
+		bpfEvent := (*tcpdropEvent)(unsafe.Pointer(&record.RawSample[0]))
+
+		reason := lookupDropReason(int(bpfEvent.Reason))
```

Comment:
IMO `lookupDropReason` should return an error and in case there is an error we should use the logger to print it instead of setting to reason.